### ISSUE TITLE
Modernize type annotations

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#7098]
After #6905, we can modernize our type annotations in various ways by doing the following:

- Use `from __future__ import annotations` ([PEP 563](https://peps.python.org/pep-0563/), [mypy docs](https://mypy.readthedocs.io/en/stable/runtime_troubles.html#future-annotations-import-pep-563)) - initial support for our introspection landed in #5769 already
- Configure `flake8-future-import` to enforce it
- Switch to [PEP 585](https://peps.python.org/pep-0585/) annotations, i.e. use builtin `tuple`, `dict`, `list`, `set` etc.
- Adopt [PEP 604](https://peps.python.org/pep-0604/), i.e. use `|` instead of `Union` and `| None` instead of `Optional`

Looks like pyupgrade and isort/reorder_python_imports have some tooling for this: [Python Type Hints - How to Upgrade Syntax with pyupgrade - Adam Johnson](https://adamj.eu/tech/2021/05/21/python-type-hints-how-to-upgrade-syntax-with-pyupgrade/)

**Comment by The-Compiler** [comment]
Looks like we can use `isort` (or `reorder-python-imports`) to add the `__future__` import for us. Thus, probably makes sense to do #1455 before this.

**Comment by The-Compiler** [comment]
Looks like @AlexWaygood has used various scripts to update the typeshed stubs:

- [Use PEP 585 syntax wherever possible by AlexWaygood · Pull Request #6717 · python/typeshed](https://github.com/python/typeshed/pull/6717)
- [Use lowercase `type` everywhere by AlexWaygood · Pull Request #6853 · python/typeshed](https://github.com/python/typeshed/pull/6853) (probably easier to do by hand for us)
- [Always use `_typeshed.Self`, where applicable](https://github.com/python/typeshed/pull/6880)
- [Fix return annotations of several methods that return `self` at runtime by AlexWaygood · Pull Request #7070 · python/typeshed](https://github.com/python/typeshed/pull/7070)
- [Use PEP 604 syntax wherever possible by AlexWaygood · Pull Request #7493 · python/typeshed](https://github.com/python/typeshed/pull/7493)
- [Use `TypeAlias` where possible for type aliases by AlexWaygood · Pull Request #7630 · python/typeshed](https://github.com/python/typeshed/pull/7630)
- [More `TypeAlias` fixes by AlexWaygood · Pull Request #7667 · python/typeshed](https://github.com/python/typeshed/pull/7667)

I wonder if we could adjust those to use them against our codebase as well (for things that pyupgrade does not handle), or perhaps even contribute those things to pyupgrade...

**Comment by AlexWaygood** [comment]
Hi! You're very welcome to reuse my scripts 🙂 (I'd appreciate some credit if you want to contribute them to somewhere like pyupgrade, though.)

A few caveats/words of caution:

- A lot of these scripts make use of the fact that `.pyi` stub files only allow a limited subset of the range of expressions that are allowed in `.py` files. For example, https://github.com/python/typeshed/pull/7667 relies on the fact that a BinOp expression on the right-hand side of an assignment can _only_ ever be a PEP-604 type alias in the context of a stub file. But in a `.py` file, the following are obviously not valid type aliases:
  ```python
  X = 1 +2
  Y = 9 | 2
  Z = 'foo' + 'bar'
  ```
- https://github.com/python/typeshed/pull/7070 executes arbitrary code by cycling through the stdlib and dynamically importing all modules. It's _relatively_ safe to do that kind of thing with the stdlib, but obviously make sure that all dangerous code is behind an `if __name__ == "__main__" guard` before you start using a script like that on your code base :)

- Basically none of these scripts have been rigorously tested, I just went with the first thing that achieved good results! (And I manually reviewed the output after the script had done its magic.) You could also check out some of the work we've been doing over at [flake8-pyi](https://github.com/PyCQA/flake8-pyi), where we've added lots of error codes to enforce modern syntax for type hints in stub files. It's a linter (it doesn't reformat your code for you), but it's better tested than the scripts I've been using to do the refactoring. A lot of the checks we've introduced there are relevant to `.py` files as well, and could potentially be added to tools like pyupgrade.

**Comment by AlexWaygood** [comment]
> I've only started to dabble into AST manipulation recently

It's _highly_ addictive once you get into it 😃

> Any idea how much work it would be to adopt a subset of them to work with code using type annotations in general rather than just pyi files?

I don't think it would be too much work at all, actually, and it's been on my mind! We're resistant to introducing complexity into flake8-pyi so that it can check `.py` files as well (see https://github.com/PyCQA/flake8-pyi/issues/118) -- we'd have to introduce `if file.extension == pyi` checks everywhere, and it would be just a little bit horrible. But if there were a way of splitting some of the checks out into a separate flake8 plugin, that could then check `.py` and `.pyi` files... that might be interesting 🙂 sadly not a top priority for me at the moment, though!

**Comment by AlexWaygood** [comment]
Here's some of the checks that could pretty easily apply to a `.py` file as well as a `.pyi` file:

- Y016 (unions shouldn't contain duplicates)
- Y018 (`TypeVar`s, `ParamSpec`s and `TypeVarTuple`s with names starting with a leading underscore should be used at least once in the file they're defined in).
- Y019 and Y034: use `Self` types where applicable.
- Y022, Y023, Y025, Y027, Y038: various imports-linting error codes. These should all be fine in a `.py` file if you're targeting 3.9+ (or if you're using `from __future__ import annotations`).
- Y024 & Y028: use class-based `typing.NamedTuple` syntax instead of assignment-based `NamedTuple` or `collections.namedtuple`.
- Y030: no more than one `Literal` member in a `Union`.
- Y031: Use class-based syntax for `TypedDict`s wherever possible.
- Y032: use `object` instead of `Any` for the second argument of `__eq__` and `__ne__` methods.
- Y033: Use PEP 526 syntax instead of type comments (should be fine as long as you're not supporting Python <3.6.
- Y036: Basic type-checking of `__(a)exit__` methods.
- Y037: only works in a `.py` file if you're targeting 3.10+.

**Comment by The-Compiler** [comment]
Once we have a nicer way of annotating `Optional[T]` (as `T | None`), we should probably also look into enabling `no_implicit_optional`.

**Comment by The-Compiler** [comment]
Also interesting: [plinss/flake8-modern-annotations: flake8 plugin to validate type annotations accoring to modern practices - Mirror of https://gitlab.linss.com/open-source/flake8/flake8-modern-annotations](https://github.com/plinss/flake8-modern-annotations)

**Comment by The-Compiler** [comment]
Nowadays ruff seems to be able to take care of most of this, via its [pyupgrade integration](https://docs.astral.sh/ruff/rules/#pyupgrade-up). I reviewed a PR for that over in pytest, and it seemed to be looking quite good:

- https://github.com/pytest-dev/pytest/pull/12467

`pyproject.toml` changes for inspiration:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 0627c94c732..734c508ca70 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -98,6 +98,7 @@ lint.select = [
     "D",       # pydocstyle
     "E",       # pycodestyle
     "F",       # pyflakes
+    "FA100",   # add future annotations
     "I",       # isort
     "PGH004",  # pygrep-hooks - Use specific rule codes when using noqa
     "PIE",     # flake8-pie
@@ -155,6 +156,10 @@ lint.per-file-ignores."src/_pytest/_version.py" = [
 lint.per-file-ignores."testing/python/approx.py" = [
     "B015",
 ]
+lint.extend-safe-fixes = [
+    "UP006",
+    "UP007",
+]
 lint.isort.combine-as-imports = true
 lint.isort.force-single-line = true
 lint.isort.force-sort-within-sections = true
@@ -164,9 +169,13 @@ lint.isort.known-local-folder = [
 ]
 lint.isort.lines-after-imports = 2
 lint.isort.order-by-type = false
+lint.isort.required-imports = [
+    "from __future__ import annotations",
+]
 # In order to be able to format for 88 char in ruff format
 lint.pycodestyle.max-line-length = 120
 lint.pydocstyle.convention = "pep257"
+lint.pyupgrade.keep-runtime-typing = false
 
 [tool.pylint.main]
 # Maximum number of characters on a single line.

```

**Comment by The-Compiler** [comment]
Now that we dropped Python 3.8 support (#8335), we switched to using built-in generics ([PEP-585](https://peps.python.org/pep-0585/)). Finally `dict[str, int]`, `list[str]`, etc.!

**Comment by The-Compiler** [comment]
Wayyy later than I would have liked, I now modernized most usages of `Union` and `Optional` along with dropping Python 3.9 support.

A few loose ends to look at still:

- Some more `Union` and `Optional` still exist with string annotations. We should introduce `from __future__ import annotations` everywhere and then also modernize those.
- I haven't looked at `typing.Self` yet, probably that makes sense to use in some places we have string annotations now.
- Possibly more from https://github.com/qutebrowser/qutebrowser/issues/7098#issuecomment-1109959035
- We still should actually enforce all this via a linter, probably best done after switching to ruff (#7468)
